components/api/src/main/java/com/hotels/styx/api/Eventual.java

```diff
@@ -113,6 +114,26 @@ public Eventual(Publisher<T> publisher) {
                 .onErrorResume(value -> Mono.from(errorHandler.apply(value))));
     }
 
+    /**
+     * Add a behavior that is triggered when this eventual emits an item successfully.
+     *
+     * @param emitted - the callback to be executed when an item is emitted
+     * @return a new {@link Eventual} with the callback registered
+     */
+    public Eventual<T> doOnEmit(Consumer<T> emitted) {
+        return fromMono(Mono.from(publisher).doOnNext(emitted));
+    }
+
+    /**
+     * Add a behavior that is triggered when this eventual terminates with an error.
+     *
+     * @param emitted - a callback to be executed when the  {@link Throwable} that caused the termination
+     * @return a new {@link Eventual} with the callback registered
+     */
+    public Eventual<T> doOnError(Consumer<? super Throwable> emitted) {
+        return fromMono(Mono.from(publisher).doOnError(emitted));
+    }
```

Comment:
This would catch errors that occur *before* the HTTP headers are received. It won't catch any errors that happen while the content is streaming through.

Reply:
Yeah, so for our case since we use a doOnNext of the mono, and then we setup a consumer<Signal<ByteBuff>> for the byteBuffer when we do receive a LiveHttpResponse. The only thing we do here is to grab the number of bytes that were streamed, so onError here is a no-op from our perspective as well.

Reply:
Right, so this would catch 5xx etc. generated by styx proxy, this is what we have been using till now. If content streaming is terminated prematurely does that result a malformed response? something that would be processed by proxies and ELBs above CGP as an error and would be displayed to users? This is something we haven't been able to get clarity on.

Reply:
> If content streaming is terminated prematurely does that result a malformed response?

It would result in a terminated TCP connection.

Styx processes all HTTP events immediately as they occur. When response headers arrive from origins, they traverse as `LiveHttpResponse` over the `Eventual` chain, and the headers are sent to the requestor straight away.

At that point, styx (HttpPipelineHandler) subscribes to the `LiveHttpResponse.body()`. Not all body data has necessarily even arrived at styx at this point: The origin might not yet have even started to write the response data. All data chunks will be processed when they arrive.

As a result, the downstream requestor might have received the response headers while the origin is still sending the response body data.

So, in face of data processing failures, the following outcomes are possible:

1) The downstream requestor has received the response headers (eg 200 OK) and possibly some data. But the TCP socket is abruptly terminated (TCP reset, or a sudden FIN from Styx).

2) The downstream requestor don't see the headers, but the TCP socket is abruptly closed with TCP reset or a FIN.

Styx has no other choice but to drop the connection. It is too late to retrospectively send an error code.

Reply:
Thank you. Got it, so I do agree to be explicit is better, so the change here is kind of misleading in terms of what an error is. Also, it seems good to separate them out, do you think we could add logic to the ResponseEventListener to add the side effects then? We are still opposed to using the map operator, mainly because of its intent of transforming items vs triggering behavior. If you agree, then we could submit a PR for changes to ResponseEventListener to the main styx repo. Let us know, and thanks for the walkthrough here helps clarify some things that have been not quite clear for us.

Reply:
Hi yes, I would definitely add it in the ResponseEventListener instead. That would make much more sense to the API consumers. Also it might help to hide any changes to the underlying API.

But we should think about adding a facade to hide the response body in the consumer callbacks.